File: tests/cpp/test_null_visualizer.cpp

#include <loco.h>
#include <gtest/gtest.h>

TEST( TestLocoNullVisualizer, TestNullVisualizerFunctionality )
{
    auto vis_data = loco::TVisualData();
    vis_data.type = loco::eShapeType::CAPSULE;
    vis_data.size = { 0.1, 0.2, 0.1 };
    auto col_data = loco::TCollisionData();
    col_data.type = loco::eShapeType::CAPSULE;
    col_data.size = { 0.1, 0.2, 0.1 };

    auto body_data = loco::TBodyData();
    body_data.dyntype = loco::eDynamicsType::DYNAMIC;
    body_data.collision = col_data;
    body_data.visual = vis_data;

    auto body_obj = std::make_unique<loco::TSingleBody>( "body_0", body_data, tinymath::Vector3f( 1.0, 1.0, 1.0 ), tinymath::Matrix3f() );
    auto scenario = std::make_unique<loco::TScenario>();
    scenario->AddSingleBody( std::move( body_obj ) );

    auto visualizer = std::make_unique<loco::TNullVisualizer>( scenario.get(), 1024, 768, true, false );
    auto cameraRef = visualizer->CreateCamera( "cam_orbit_0", 
                                               loco::eVizCameraType::ORBIT,
                                               { 3.0f, 3.0f, 3.0f },
                                               { 0.0f, 0.0f, 0.0f } );
    auto lightRef = visualizer->CreateLight( "light_point_0",
                                             loco::eVizLightType::POINT,
                                             { 0.4f, 0.4f, 0.4f },
                                             { 0.8f, 0.8f, 0.8f },
                                             { 0.8f, 0.8f, 0.8f } );
    visualizer->Initialize();
    visualizer->Update();
    visualizer->Reset();

    cameraRef->SetPosition( { 5.0f, 5.0f, 5.0f } );
    cameraRef->SetTarget( { 0.0f, 0.0f, 1.0f } );

    lightRef->SetPosition( { 0.0f, 0.0f, 7.0f } );
    lightRef->ChangeIntensity( 0.9f );

    EXPECT_FALSE( visualizer->HasCameraNamed( "foooooooo" ) );
    EXPECT_FALSE( visualizer->HasLightNamed( "fuuuuuuuu" ) );
    EXPECT_TRUE( visualizer->HasCameraNamed( "cam_orbit_0" ) );
    EXPECT_TRUE( visualizer->HasLightNamed( "light_point_0" ) );
    EXPECT_TRUE( visualizer->GetCameraByName( "cam_orbit_0" ) != nullptr );
    EXPECT_TRUE( visualizer->GetLightByName( "light_point_0" ) != nullptr );
    EXPECT_TRUE( tinymath::allclose( visualizer->GetCameraByName( "cam_orbit_0" )->position(), cameraRef->position() ) );
    EXPECT_TRUE( tinymath::allclose( visualizer->GetLightByName( "light_point_0" )->ambient(), lightRef->ambient() ) );

    EXPECT_EQ( cameraRef->type(), loco::eVizCameraType::ORBIT );
    EXPECT_TRUE( tinymath::allclose( cameraRef->position(), { 5.0f, 5.0f, 5.0f } ) );
    EXPECT_TRUE( tinymath::allclose( cameraRef->target(), { 0.0f, 0.0f, 1.0f } ) );

    EXPECT_EQ( lightRef->type(), loco::eVizLightType::POINT );
    EXPECT_TRUE( tinymath::allclose( lightRef->position(), { 0.0f, 0.0f, 7.0f } ) );
    EXPECT_TRUE( tinymath::allclose( lightRef->ambient(), { 0.4f, 0.4f, 0.4f } ) );
    EXPECT_TRUE( tinymath::allclose( lightRef->diffuse(), { 0.8f, 0.8f, 0.8f } ) );
    EXPECT_TRUE( tinymath::allclose( lightRef->specular(), { 0.8f, 0.8f, 0.8f } ) );
    EXPECT_TRUE( std::abs( lightRef->intensity() - 0.9f ) < 1e-6 );
}

File: legacy/tests/python/test_null_visualizer.py
#!/usr/bin/env python

import loco
import tinymath as tm
import numpy as np
import gc

def test_null_visualizer_functionality() :
    vis_data = loco.sim.VisualData()
    vis_data.type = loco.sim.ShapeType.CAPSULE
    vis_data.size = [ 0.1, 0.2, 0.1 ]
    col_data = loco.sim.CollisionData()
    col_data.type = loco.sim.ShapeType.CAPSULE
    col_data.size = [ 0.1, 0.2, 0.1 ]

    body_data = loco.sim.BodyData()
    body_data.dyntype = loco.sim.DynamicsType.DYNAMIC
    body_data.collision = col_data
    body_data.visual = vis_data

    body_obj = loco.sim.SingleBody( 'body_0', body_data, [ 1.0, 1.0, 1.0 ], np.identity( 3 ) )
    scenario = loco.sim.Scenario()
    scenario.AddSingleBody( body_obj )

    visualizer = loco.sim.NullVisualizer( scenario )
    assert ( visualizer.GetCameraByName( 'cam_orbit_0' ) == None )
    assert ( visualizer.GetLightByName( 'light_point_0' ) == None )

    camera = visualizer.CreateCamera( 'cam_orbit_0',
                                      loco.sim.VizCameraType.ORBIT,
                                      [ 3.0, 3.0, 3.0 ],
                                      [ 0.0, 0.0, 0.0 ] )
    light = visualizer.CreateLight( 'light_point_0',
                                    loco.sim.VizLightType.POINT,
                                    [ 0.4, 0.4, 0.4 ],
                                    [ 0.8, 0.8, 0.8 ],
                                    [ 0.8, 0.8, 0.8 ] )
    visualizer.Initialize()
    visualizer.Update()
    visualizer.Reset()

    camera.position = [ 5.0, 5.0, 5.0 ]
    camera.target = [ 0.0, 0.0, 1.0 ]

    light.position = [ 0.0, 0.0, 7.0 ]
    light.intensity = 0.9

    assert ( visualizer.HasCameraNamed( 'cam_orbit_0' ) == True )
    assert ( visualizer.HasLightNamed( 'light_point_0' ) == True )
    assert ( visualizer.GetCameraByName( 'cam_orbit_0' ) != None )
    assert ( visualizer.GetLightByName( 'light_point_0' ) != None )
    assert np.allclose( visualizer.GetCameraByName( 'cam_orbit_0' ).position, camera.position )
    assert np.allclose( visualizer.GetLightByName( 'light_point_0' ).ambient, light.ambient )

    assert ( camera.type == loco.sim.VizCameraType.ORBIT )
    assert np.allclose( camera.position, [ 5.0, 5.0, 5.0 ] )
    assert np.allclose( camera.target, [ 0.0, 0.0, 1.0 ] )

    assert ( light.type == loco.sim.VizLightType.POINT )
    assert np.allclose( light.position, [ 0.0, 0.0, 7.0 ] )
    assert np.allclose( light.ambient, [ 0.4, 0.4, 0.4 ] )
    assert np.allclose( light.diffuse, [ 0.8, 0.8, 0.8 ] )
    assert np.allclose( light.specular, [ 0.8, 0.8, 0.8 ] )
    assert np.allclose( light.intensity, 0.9 )
    del visualizer
    del scenario

if __name__ == '__main__' :
    _ = input( 'Press ENTER to start test : test_null_visualizer_functionality' )
    test_null_visualizer_functionality()

    _ = input( 'Press ENTER to continue ...' )